core/jobs/batch_jobs/exp_version_history_computation_job.py

```diff
@@ -0,0 +1,562 @@
+                exp_models.ExplorationCommitLogEntryModel.get_all(
+                    include_deleted=False
+                )
+            )
+            | 'Create key-value pairs with id and commit log models' >>
+                beam.Map(lambda model: (model.exploration_id, model))
+        )
+
+        all_version_history_models = (
+            all_commit_logs
+            | 'Get already existing ExplorationVersionHistoryModels' >>
+                ndb_io.GetModels(
+                    exp_models.ExplorationVersionHistoryModel.get_all(
+                        include_deleted=False
+                    )
+                )
+            | 'Create key-value pairs with id and version history models' >>
+                beam.Map(lambda model: (model.exploration_id, model))
+        )
+
+        model_groups = (
+            ({
+                'exp_models_v1': all_explorations_v1,
+                'exp_models_vlatest': all_explorations_vlatest,
+                'commit_log_models': all_commit_logs,
+                'version_history_models': all_version_history_models
+            })
+            | 'Group by key' >> beam.CoGroupByKey()
+            | 'Get rid of exploration id' >>
+                beam.Values() # pylint: disable=no-value-for-parameter
+            | 'Filter valid model groups' >> beam.Filter(
+                self._filter_valid_model_group
+            )
+            | 'Format valid model groups' >> beam.Map(
+                self._convert_to_formatted_model_group_dict
+            )
+        )
+
+        version_history_models = (
+            model_groups
+            | 'Create the version history models for each valid exploration' >>
+                beam.Map(self._create_version_history_models)
+            | 'Flatten the models' >> beam.FlatMap(lambda x: x)
+        )
+
+        unused_put_result = (
+            version_history_models
+            | 'Save the models to the datastore' >> ndb_io.PutModels()
+        )
+
+        report_number_of_exps_queried = (
+            all_explorations_vlatest
+            | 'Count queried explorations' >>
+                job_result_transforms.CountObjectsToJobRunResult('ALL EXPS')
+        )
+
+        report_number_of_valid_exps_queried = (
+            all_explorations_v1
+            | 'Count valid queried explorations' >>
+                job_result_transforms.CountObjectsToJobRunResult(
+                    'ALL VALID EXPS'
+                )
+        )
+
+        report_number_of_models_created = (
+            version_history_models
+            | 'Count number of models created' >>
+                job_result_transforms.CountObjectsToJobRunResult(
+                    'CREATED VERSION HISTORY MODELS'
+                )
+        )
+
+        return (
+            (
+                report_number_of_exps_queried,
+                report_number_of_valid_exps_queried,
+                report_number_of_models_created
```

Comment:
Consider having better outputs to easily validate or invalidate a run. For example, if you output the number of version history models created, the number of existing version history models and the total number of commit log entries, it should be possible to validate that the counts add up as expected. Or that the things marked as invalid should be sent back to you (maybe the ids or something) so that you can investigate them further.  The more debugging information you get out of the job, the better :)

Reply:
Improved the output now.